src/ctr_logging_buffer.c

```diff
@@ -0,0 +1,482 @@
+
+	/* Publish the buffer with release semantics */
+	atomic_store_explicit(&g_log_buffer, new_buffer, memory_order_release);
+
+	exit_critical_section(&old_mask);
+
+	/* Setup timer outside critical section to avoid deadlock */
+	setup_flush_timer();
+
+	ninfo("Async logging initialized successfully");
+	return true;
+}
+
+void shutdown_async_logging(void)
+{
+	sigset_t old_mask;
+	if (enter_critical_section(&old_mask) != 0) {
+		return;
+	}
+
+	async_log_buffer_t *buffer = atomic_load_explicit(&g_log_buffer, memory_order_acquire);
+	if (!buffer) {
+		exit_critical_section(&old_mask);
+		return;
+	}
+
+	/* Mark shutdown to prevent new operations */
+	atomic_store_explicit(&buffer->shutdown_requested, true, memory_order_release);
+
+	/* Clean up timer first (outside critical section to avoid deadlock) */
+	exit_critical_section(&old_mask);
+	cleanup_flush_timer();
+
+	/* Re-enter critical section for final operations */
+	if (enter_critical_section(&old_mask) != 0) {
+		return;
+	}
+
+	buffer = atomic_load_explicit(&g_log_buffer, memory_order_acquire);
+	if (!buffer) {
+		exit_critical_section(&old_mask);
+		return;
+	}
+
+	/* Perform final flush */
+	flush_buffer_contents_locked();
+
+	/* Get final statistics before cleanup */
+	size_t written = buffer->entries_written;
+	size_t dropped = buffer->entries_dropped;
+
+	/* Clear the global pointer with release semantics */
+	atomic_store_explicit(&g_log_buffer, NULL, memory_order_release);
+
+	exit_critical_section(&old_mask);
+
+	ninfof("Async logging shutdown. Entries written: %zu, dropped: %zu", written, dropped);
+	free(buffer);
+}
+
+bool write_to_logs_buffered(stdpipe_t pipe, char *buf, ssize_t num_read)
+{
+	/* Validate input parameters first */
+	if (!buf || num_read < 0) {
+		return false;
+	}
+
+	sigset_t old_mask;
+	if (enter_critical_section(&old_mask) != 0) {
+		/* Failed to enter critical section, fallback to direct logging */
+		return write_to_logs(pipe, buf, num_read);
+	}
+
+	/* Handle empty writes explicitly - these are valid drain operations */
+	if (num_read == 0) {
+		exit_critical_section(&old_mask);
+		return write_to_logs(pipe, buf, num_read);
+	}
+
+	/* Check buffer state atomically in critical section */
+	async_log_buffer_t *buffer = atomic_load_explicit(&g_log_buffer, memory_order_acquire);
+	bool initialized = buffer ? atomic_load_explicit(&buffer->initialized, memory_order_acquire) : false;
+	bool shutdown = buffer ? atomic_load_explicit(&buffer->shutdown_requested, memory_order_acquire) : false;
+
+	if (!buffer || !initialized || shutdown) {
+		exit_critical_section(&old_mask);
+		return write_to_logs(pipe, buf, num_read);
+	}
+
+	/* Try to add to buffer */
+	bool success = add_entry_to_buffer_locked(pipe, buf, num_read);
+
+	if (!success) {
+		/* Buffer full - flush and try again */
+		flush_buffer_contents_locked();
+		success = add_entry_to_buffer_locked(pipe, buf, num_read);
+	}
+
+	if (!success) {
```

Comment:
This check appears to only be required to check the return value from line 204, as the value of `success` will still be true if it is true at line 201.

Is it here to save indentation?

Can we consider moving this to follow line 204 in that same block?